ndless-sdk/libndls/config.c

```diff
@@ -102,73 +141,146 @@ void cfg_open_file(const char *filepath) {
 				str = value;
 				while (*str != 0 && *str != ' ') str++;
 				if (*str == ' ') *str = 0; // remove the spaces after the value
-				kv_offsets[kv_index][0] = kv - file_content;
-				kv_offsets[kv_index++][1] = value - file_content;
+				struct cfg_entry *entry = &cfg_entries[kv_index++];
+				strlcpy(entry->key, kv, 15);
+				size_t val_len = strlen(value) + 1;
+				entry->value_sz = pow2_roundup(val_len);
+				entry->value = malloc(entry->value_sz);
+				strlcpy(entry->value, value, entry->value_sz);
+				// printf("entry %s=%s\n", entry->key, entry->value);
 				if (kv_index >= max_kv_num) break; // too many key-value pairs
 			}
 		}
 		if (!end_of_pair) break; // end of file
 		str = end_of_pair; // next key value
 		if (str >= end_of_file) break; // end of file
 		if (has_comment) {
-			str = strpbrk(str, "\n");  // skip it
+			char *ns = strpbrk(str, "\n");
+			if(value) {
+				if(!ns) ns = end_of_file;
+				char *cmt = cfg_comments[kv_index - 1] = malloc(ns - str + 1);
+				memcpy(cmt, str, ns - str);
+				cmt[ns - str] = '\0';
+			}
+			str = ns;  // skip it
 			if (!str) break; // end of file
 			str++;
 		}
 	}
-	kv_num = kv_index;
+	loop_end:
+	cmt_num = kv_num = kv_index;
+	strlcpy(open_file, filepath, FILENAME_MAX);
+	free(file_content);
 }
 
+static char cfg_path[FILENAME_MAX] = {0};
+
 static int cfg_locate_cfg_file(char *dst_path, size_t dst_path_size) {
-	return locate("ndless.cfg.tns", dst_path, dst_path_size);
+	if(cfg_path[0] == 0) snprintf(cfg_path, FILENAME_MAX, "%s%s", get_documents_dir(), "ndless/ndless.cfg.tns");
+	if(access(cfg_path, F_OK) == -1) {
+		int l = locate("ndless.cfg.tns", cfg_path, FILENAME_MAX);
+		if(l != 0) return l;
+	}
+	size_t actual_path_len = strlen(cfg_path);
+	if(actual_path_len+1 > dst_path_size) return 1;
+	memcpy(dst_path, cfg_path, actual_path_len);
+	dst_path[actual_path_len] = 0;
+	return 0;
 }
 
 void cfg_open(void) {
-	char path[300];
+	char path[FILENAME_MAX];
```

Comment:
PATH_MAX?

Reply:
I saw FILENAME_MAX used in this way in a few other places in ndless (i.e. in `ld_exec_with_args`). Doesn't make it right, but that was my logic.